### PR TITLE
[DashboardLayout] Support route patterns in navigation

### PR DESCRIPTION
**Pull request by apedroferreira**
Closes https://github.com/mui/mui-toolpad/issues/3974 (partially, must create new partial issue for `PageContainer` component and follow https://github.com/mui/mui-toolpad/pull/3991#issuecomment-2310618664 + https://github.com/mui/mui-toolpad/pull/3991#pullrequestreview-2262532367

Adds an optional `pattern` property to navigation page items so that these can be marked as selected according to that pattern (using [path-to-regexp](https://www.npmjs.com/package/path-to-regexp)), instead of the default behavior that uses their `segment` property.

- [x] Tests
- [x] Documentation (no demo as this doesn't make a lot of sense to demo, just an explanation)

https://deploy-preview-3991--mui-toolpad-docs.netlify.app/toolpad/core/react-dashboard-layout/#navigation-pattern-matching

Also simplifies the logic for the console warning in `DashboardLayout` for duplicate navigation items.

No changes to `PageContainer` or any more complex behavior for now, those will be done separately, as well as refactored logic for the selected item.

### CODE REVIEW

**Comment by apedroferreira** [comment]
In a further PR, will add further refactoring to keep track of one only selected list item and share that logic between `DashboardLayout` and `PageContainer` components.

File: docs/data/toolpad/core/components/dashboard-layout/dashboard-layout.md [hunk 1]
```diff
@@ -78,10 +78,20 @@ Nested navigation structures can be placed in the sidebar as items with the form
...
 {{"demo": "DashboardLayoutNavigationActions.js", "height": 400, "iframe": true}}
 
+### Navigation Pattern Matching
+
+Navigation links have an optional `pattern` prop to define a pattern to be matched for the item to be marked as selected.
+Any path supported by the [path-to-regexp](https://www.npmjs.com/package/path-to-regexp) library can be used, such as:
+
+- Standard paths: `/foo`
+- Named parameters: `/:foo/:bar`
+- Unnamed parameters: `/:foo/(.*)`
```
Comment (Janpot) [inline docs/data/toolpad/core/components/dashboard-layout/dashboard-layout.md:92]: I don't think named vs unnamed is relevant to us. I'd just say "Parameters" on the line above. I'd also use something slightly more real-life as an example, like "orders"

 ```md
 - Constant path: `/orders`
 - Dynamic segment: `/orders/:segment`
 - Zero or more segments: `/orders{/:segment}*`
 - One or more segments: `/orders{/:segment}+`
 - Optional segment: `/orders{/:segment}+?`
```

File: docs/data/toolpad/core/components/dashboard-layout/dashboard-layout.md [hunk 2]
```diff
@@ -78,10 +78,20 @@ Nested navigation structures can be placed in the sidebar as items with the form
...
 
-Navigation links have an optional `action` prop that can be used to render any content on the right-side of the respective list item, such as badges with numbers, or buttons to toggle a popover menu.
+Navigation links have an optional `action` prop to render any content on the right-side of the respective list item, such as badges with numbers, or buttons to toggle a popover menu.
 
 {{"demo": "DashboardLayoutNavigationActions.js", "height": 400, "iframe": true}}
 
+### Navigation Pattern Matching
+
+Navigation links have an optional `pattern` prop to define a pattern to be matched for the item to be marked as selected.
+Any path supported by the [path-to-regexp](https://www.npmjs.com/package/path-to-regexp) library can be used, such as:
```
Comment (Janpot) [inline docs/data/toolpad/core/components/dashboard-layout/dashboard-layout.md:88]: I wouldn't encourage users too much 😄 
```suggestion
This feature is built on top of the [path-to-regexp](https://www.npmjs.com/package/path-to-regexp) library.
```

File: docs/data/toolpad/core/components/dashboard-layout/dashboard-layout.md [hunk 3]
```diff
@@ -78,10 +78,20 @@ Nested navigation structures can be placed in the sidebar as items with the form
...
+
+- Standard paths: `/foo`
+- Named parameters: `/:foo/:bar`
+- Unnamed parameters: `/:foo/(.*)`
+- Optional parameters: `/:foo{/:bar}?`
```
Comment (Janpot) [inline docs/data/toolpad/core/components/dashboard-layout/dashboard-layout.md:93]: We could maybe add demo. a "product" page that has links to show the "next product" and the current path. e.g.

```
current path: /products/20
title: Item 20
[next product](/products/21)
```

**Review by Janpot — APPROVED** [review]
Looks good 👍

**Pull request merged**